### ISSUE TITLE
SignFile cannot find signtool.exe with MSBuild v17.1 installed from Visual studio build tools 2022

### ISSUE THREAD

**Issue by dqwork** [dotnet/msbuild#7616]
### Issue Description
MSBuild Version: 17.1.0+ae57d105c for .NET Framework
dotnet version: 6.0.202

Visual studio build tools 2022 installed with following Components/Workloads

Microsoft.VisualStudio.Workload.ManagedDesktopBuildTools
Microsoft.VisualStudio.Workload.MSBuildTools
Microsoft.VisualStudio.Workload.WebBuildTools
Microsoft.VisualStudio.Component.Windows10SDK.20348

Signtool.exe exists in the following places

C:\Program Files (x86)\Microsoft SDKs\ClickOnce\SignTool\signtool.exe
C:\Program Files (x86)\Windows Kits\10\bin\10.0.20348.0\arm\signtool.exe
C:\Program Files (x86)\Windows Kits\10\bin\10.0.20348.0\arm64\signtool.exe
C:\Program Files (x86)\Windows Kits\10\bin\10.0.20348.0\x64\signtool.exe
C:\Program Files (x86)\Windows Kits\10\bin\10.0.20348.0\x86\signtool.exe

Attempting to Sign a clickonce executable using MSBuild SignFile Task in our build pipeline fails

error MSB3482: An error occurred while signing:
 SignTool.exe not found

### Steps to Reproduce

I can't share too much as I'm working on internal projects. If more info is needed I'm happy to contract directly and share information directly with microsoft (rather than on a public forum)

### Expected Behavior
The executable is succesfully signed

### Actual Behavior
this error is returned

error MSB3482: An error occurred while signing:
 SignTool.exe not found

### Analysis
On an older server that is running our build pipeline, that has VS Build toos 2017 and 2019 installed this runs fine. it has older Windows Kits installed (that now seem to be unavailable)

** Old server that work's list of SignTool.exe locations **
C:\Program Files\Microsoft SDKs\Windows\v6.0A\Bin\signtool.exe
C:\Program Files (x86)\Microsoft SDKs\ClickOnce\SignTool\signtool.exe
C:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\Bin\signtool.exe
C:\Program Files (x86)\Microsoft SDKs\Windows\v7.1A\Bin\signtool.exe
C:\Program Files (x86)\Windows Kits\8.0\bin\x64\signtool.exe
C:\Program Files (x86)\Windows Kits\8.0\bin\x86\signtool.exe

I'm wondering if for some reason the task is looking in a 'old' location and maybe shouldn't be? 

### Versions & Configurations
MSBuild Version: 17.1.0+ae57d105c for .NET Framework

Windows Server 2019 

-->

**Comment by rainersigwald** [comment]
@sujitnayak can you take a look at this?

**Comment by sujitnayak** [comment]
@rainersigwald  Did this PR make it to 17.1? 
https://github.com/dotnet/msbuild/pull/6463

@dqwork Does this regkey exist?
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\ClickOnce\SignTool and does the Path value point to a folder with signtool.exe?

Thanks.

**Comment by sujitnayak** [comment]
@dqwork Please go to Help -> About Visual Studio in VS and paste the full Version info. Thanks.

**Comment by dqwork** [comment]
@sujitnayak - Thanks for the quick repsonses.

It response to your first question. The reg Key does exist
![image](https://user-images.githubusercontent.com/77854269/168260696-4e9f69ef-14bc-4910-98c0-6604e888e337.png)
and the folder path does contain signtool.exe
![image](https://user-images.githubusercontent.com/77854269/168260782-c6399372-a53c-4c89-a433-ce4829ab28e1.png)


In terms of getting the full version info - I'm not sure how to do that when we only have build tools installed - theres no UI for me to click 'Help'? In the Visual studio installer is says we have version 17.1.6 of Build tools installed

**Comment by dqwork** [comment]
I've also tried updating to 17.2.0 of VS Build tools and continue to have the same issue

**Comment by sujitnayak** [comment]
@dqwork  thanks for that info. Is it possible for you to install the corresponding VS SKU on this machine and try to see if a simple WinForms app can be signed and run on this machine? I am trying to figure out if this issue is isolated to the Build SKU
Thanks.

**Comment by dqwork** [comment]
Hi @sujitnayak - I installed Visual Studio (full IDE) with the desktop development workload and Clickonce publishing.

From within Visual studio I created a simple winforms app and set the properties up to sign the clickonce manifest and published it to a local folder. This worked fine and when I attempted to install the manifest it was properly sign.

On the same machine I used MSBuild from the command line to try and assign the original file I wanted to sign. This still fails with the same error. 

To summarize: Using Visual Studio 2022 to sign the manifest works fine. Using MSBuild (first installed from VS Build tools 2022) does not work and fails with SignTool.exe not found - even when Full visual studio is installed

**Comment by sujitnayak** [comment]
@dqwork Can you paste the command line you're using that is failing?
thanks.

**Comment by dqwork** [comment]
running in this directory C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\MSBuild\Current\Bin

.\msbuild.exe <msbuild file> /property:SignTargetPath=<.exe that its signing> /property:CertificatePath=<path to pfx file> /property:CertificatePassword=<pfx file password> /property:TimestampUrl=<timestamp Url> /ToolsVersion:4.0

**Comment by dqwork** [comment]
Interestingly if I omit '/ToolsVersion:4.0' it finds the tool and runs fine

**Comment by dqwork** [comment]
if I change it to /ToolsVersion:4.5 - I'm told the only valid options are 'Current' or '4.0' - however as we've been discussing, '4.0' causes it to fail

**Comment by dqwork** [comment]
Hi @sujitnayak,

So we are using our own MSBuild file that does a couple of things before calling SignFile task.

If you'd like to contact me directly via email I can share that file with you, but for now I'll give you a breakdown of what it does and how its calling the SignFile task.

First it loads some MSBuild tasks from a nuget file, and uses one of those Targets to add a pfx file to the certificate store and that returns a thumb print. 

Finally it calls into SignFile like this
where CertificateThumbprint is returned from the task to add the cert
SignTargetPath and TimestampUrl are passed in by the caller 
TargetFrameworkVersion set to v4.0 to allow SHA1 certificates to support older clients.

```
<SignFile
            CertificateThumbprint="$(CertificateThumbprint)"
            SigningTarget="$(SignTargetPath)"
            TargetFrameworkVersion="v4.0"
            TimestampUrl="$(TimestampUrl)" />
```

**Comment by dqwork** [comment]
@sujitnayak  I've been playing around with our build pipeline and it looks like its doing something rather funky. it seems to be doing something that means when its running our custom task to install the cert and sign it, its ends up running MSBuild from here C:\Windows\Microsoft.NET\Framework\v4.0.30319\msbuild.exe - which may go someway to explaining whats happening. 

I'm going to discuss with the rest of our team to find out why this is, but I wanted to let you know before you spend anymore time investigating. I'll update with more info after I've spoken to them

**Comment by dqwork** [comment]
@sujitnayak  thanks again for taking time to help me out here. I've managed to narrow the issue and it was some odd steps in our build pipeline that were forcing it to use the old msbuild version. I've fixed that and its solved my issue
